cpp/src/jit/span.cuh

```diff
@@ -149,6 +154,9 @@ struct device_optional_span : device_span<T> {
     return !is_valid(element_index);
   }
 
+  /// @brief converts the optional span to a regular non-nullable span.
+  [[nodiscard]] __device__ base to_span() const noexcept { return static_cast<base const&>(*this); }
```

Comment:
Why are we using inheritance rather than composition here? In other words, why do we have a `device_optional_span<T>` is-a `device_span<T>` rather than has-a `device_span<T>` as a member?

Reply:
it's composition via data inheritance. it isn't exposing any of `device_span`'s methods or having the user interface with it via the `device_span` class.
Inheritance here is an implementation detail and it's only intended to make implicit conversion to `device_span` easy.